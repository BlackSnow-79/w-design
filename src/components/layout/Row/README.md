Wrapper Component around Columns

| Props          | type   | default | description                |
| -------------- | ------ | ------- | -------------------------- |
| flexWrap       | string | wrap    | flex-wrap definition       |
| flexDirection  | string | wrap    | flex-wrap definition       |
| alignItems     | string | wrap    | flex-wrap definition       |
| justifyContent | string | wrap    | flex-wrap definition       |
| noGap          | bool   | wrap    | flex-wrap definition       |
| oneline        | bool   | wrap    | Set grid columns to online |

### Default Grid

```js
<Provider>
  <Container>
    <Row>
      <Column width={1} bg="N2">
        width 1
      </Column>

      <Column width={1 / 2} bg="N3">
        width 1/2
      </Column>

      <Column width={1 / 4} bg="N4">
        width 1/4
      </Column>
      <Column width={1 / 3} bg="N2">
        width 1/3
      </Column>
    </Row>
  </Container>
</Provider>
```

### Gapless Grid

```js
<Provider>
  <Container>
    <Row gapless>
      <Column width={1} bg="N2">
        width 1
      </Column>

      <Column width={1 / 2} bg="N3">
        width 1/2
      </Column>

      <Column width={1 / 4} bg="N4">
        width 1/4
      </Column>
      <Column width={1 / 3} bg="N2">
        width 1/3
      </Column>
    </Row>
  </Container>
</Provider>
```

### Oneline Grid

Grid `oneline` prop make all child columns positioned in the same single row.

```js
<Provider>
  <Container>
    <Row oneline>
      <Column width={1} bg="N2">
        width 1
      </Column>

      <Column width={1 / 2} bg="N3">
        width 1/2
      </Column>

      <Column width={1 / 4} bg="N4">
        width 1/4
      </Column>
      <Column width={1 / 3} bg="N2">
        width 1/3
      </Column>
    </Row>
  </Container>
</Provider>
```

### Custom Grid

```js
<Provider>
  <Container>
    <Row justifyContent="space-between">
      <Column width={1} bg="N2">
        width 1
      </Column>

      <Column width={1 / 2} bg="N3" m="auto">
        width 1/2, m="auto"
      </Column>

      <Column width={1 / 4} bg="N4">
        width 1/4
      </Column>
      <Column width={1 / 3} bg="N2">
        width 1/3
      </Column>
    </Row>
  </Container>
</Provider>
```

### Responsive Column

The `width` prop will scale depend on breakpoints which have [24, 32, 48, 64, 80] in default theme

example

width={[
1, // 100% below the smallest breakpoint
1 / 2, // 50% from the next breakpoint and up
1 / 4 // 25% from the next breakpoint and up
]}

```js
<Provider>
  <Container>
    <Row>
      <Column width={[1, 1 / 2, 1 / 3, 1 / 4]} bg="N2">
        Column 1
      </Column>

      <Column width={[1, 1 / 2, 1 / 3, 1 / 4, 1]} bg="N3">
        Column 2
      </Column>
    </Row>
  </Container>
</Provider>
```

### Responsive Hidden

```js
<Provider>
  <Container>
    <Row>
      <Column width={1 / 2} bg="N2">
        Column 1
      </Column>
      <Column width={1 / 2} bg="N3">
        Column 1
      </Column>
    </Row>
  </Container>
</Provider>
```
