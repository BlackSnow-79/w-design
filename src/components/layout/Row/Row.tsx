import * as React from "react";
import {
  flexWrap,
  flexDirection,
  alignItems,
  justifyContent
} from "styled-system";
import styled from "../../../utils/styled";

export interface Props {
  gapless?: boolean;
  oneline?: boolean;
}

export interface Props {
  /** children components*/
  children?: React.ReactChild;
}

const Base = styled.div`
  display: flex;
  flex-wrap: wrap;
  box-sizing: border-box;
  margin-left: -0.4rem;
  margin-right: -0.4rem;

  ${flexWrap}
  ${flexDirection}
  ${alignItems}
  ${justifyContent}

  ${(props: Props) =>
    /* set margin left and right, and child padding to 0*/
    props.gapless &&
    ` margin-left: 0 !important;
      margin-right: 0 !important;
      * {
        padding-left: 0 !important;
        padding-right: 0 !important;
      }
  `};

  ${(props: Props) =>
    props.oneline &&
    `
      flex-wrap: nowrap !important;
      overflow-x: auto !important;
  `};
`;

const Row: React.StatelessComponent<Props> = props => <Base {...props} />;

export default Row;
