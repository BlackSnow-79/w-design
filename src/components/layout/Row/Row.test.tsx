import * as React from "react";
import { shallow } from "enzyme";
import "jest-styled-components";
import Row from "./Row";

describe("<GridWrap/>", () => {
  // Render test
  it("should renders correctly", () => {
    const wrapper = shallow(<Row />);
    expect(wrapper).toMatchSnapshot();
  });

  // Props test
  it("should has gapless props", () => {
    const wrapper = shallow(<Row />);
    wrapper.setProps({ gapless: true });
    expect(wrapper).toMatchSnapshot();
  });

  it("should has oneline props", () => {
    const wrapper = shallow(<Row />);
    wrapper.setProps({ oneline: true });
    expect(wrapper).toMatchSnapshot();
  });
});
