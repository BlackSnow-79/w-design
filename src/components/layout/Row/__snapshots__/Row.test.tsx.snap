// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridWrap/> should has gapless props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Row
    gapless={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "gapless": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "gapless": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<GridWrap/> should has oneline props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Row
    oneline={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "oneline": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "oneline": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<GridWrap/> should renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Row />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
